### PR TITLE
Function hook

### PR DESCRIPTION
**Pull request by delta2323**
As we have not fully discussed the specification of examples of `FunctionHook`s (especially `MemoryHook`), we separate #933 into two PRs, `FunctionHook` mechanism and examples of `FunctionHook` (subclasses of `FunctionHook`s).

This PR implements the former one.

We change the implementation from #933 in two points.
* Remove `preprocess`, `postprocess`, and `__call__` methods of `FunctionHook`.
* Make `chainer.global_function_hook` thread safe.

### CODE REVIEW

File: chainer/function.py [hunk 1]
```diff
@@ -135,6 +144,26 @@ def __call__(self, *inputs):
...
+
+        Contrary to ``chainer.thread_local.function_hooks``,
+        which registers its elements to all functions,
+        Function hooks in this property is specific to this function.
+        """
+        if not hasattr(self, '_local_function_hooks'):
+            self._local_function_hooks = collections.OrderedDict()
+        return self._local_function_hooks
+
+    @local_function_hooks.setter
```
Comment (unnonouno) [inline chainer/function.py]: Is this method required?

File: chainer/function.py [hunk 2]
```diff
@@ -135,6 +144,26 @@ def __call__(self, *inputs):
...
+    def local_function_hooks(self, val):
+        self._local_function_hooks = val
+
+    @local_function_hooks.deleter
```
Comment (unnonouno) [inline chainer/function.py]: Is this method required?

File: chainer/function.py [hunk 3]
```diff
@@ -135,6 +144,26 @@ def __call__(self, *inputs):
...
+    def local_function_hooks(self):
```
Comment (unnonouno) [inline chainer/function.py]: Is this mistake?

File: chainer/variable.py [hunk 4]
```diff
@@ -336,9 +337,15 @@ def add_cand(cand):
 
             in_data = tuple(x.data for x in func.inputs)
             out_grad = tuple(None if y is None else y.grad for y in outputs)
+            hooks = collections.OrderedDict(chainer.get_function_hooks())
+            hooks.update(func.local_function_hooks)
+            for hook in hooks.values():
```
Comment (unnonouno) [inline chainer/variable.py]: Use `six.itervalues`

File: chainer/variable.py [hunk 5]
```diff
@@ -336,9 +337,15 @@ def add_cand(cand):
...
+                hook.backward_preprocess(func, in_data, out_grad)
             with cuda.get_device(*(in_data + out_grad)):
                 gxs = func.backward(in_data, out_grad)
             assert len(gxs) == len(in_data)
+            for hook in hooks.values():
```
Comment (unnonouno) [inline chainer/variable.py]: Use `six.itervalues`

File: chainer/function.py (hunk 3 above)
Comment (delta2323) [inline chainer/function.py]: I fixed it.

File: tests/chainer_tests/function_hooks_tests/test_timer.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+import chainer
+from chainer import cuda
+from chainer import function_hooks
+from chainer import functions
+from chainer.functions.connection import linear
+from chainer import links
+from chainer import testing
+from chainer.testing import attr
+import numpy
```
Comment (unnonouno) [inline tests/chainer_tests/function_hooks_tests/test_timer.py]: Fix import order

File: chainer/function.py [hunk 7]
```diff
@@ -317,3 +339,178 @@ def unchain(self):
...
+class FunctionHook(object):
+    """Base class of hooks for Functions.
+
+    :class:`~chainer.function.FunctionsHook` is an callback object
+    that is registered to :class:`~chainer.function.Function`.
+    Registered function hooks are invoked before and after
+    forward and backward operations of each function.
+
+    Function hooks that derive :class:`FunctionHook` are required
+    to implement four functions:
```
Comment (unnonouno) [inline chainer/function.py]: functions -> methods ?

File: chainer/function.py [hunk 8]
```diff
@@ -317,3 +339,178 @@ def unchain(self):
...
+    of all function hooks registered to the function which holds this variable
+    as a gradient are called before (resp. after) backward propagation.
+
+    There are two ways to register :class:`~chainer.function.FunctionHook`
+    objects to :class:`chainer.function.Function` objects.
+    First one is to add the :class:`~chainer.function.FunctionHook`
+    to ``chainer.thread_local.function_hooks``.
+    This is a thread local dictionary whose keys are strings that represent
+    the names of function hooks.
+    Function hooks in ``chainer.thread_local.function_hooks`` are regared
```
Comment (unnonouno) [inline chainer/function.py]: regared -> regarded ?

File: chainer/function.py [hunk 9]
```diff
@@ -317,3 +339,178 @@ def unchain(self):
...
+
+    >>> class Model(chainer.Chain):
+    ...     def __call__(x):
+    ...         x = self.l(x)
+    ...         return F.exp(x)
+    ...
+    ... model1 = chainer.Model(l=L.Linear(10, 10))
+    ... model2 = chainer.Model(l=L.Linear(10, 10))
+    ... x = chainer.Variable(...)
+    ... with TimerHook():
```
Comment (unnonouno) [inline chainer/function.py]: We need to store this `TimerHook` instance to show elapsed time?

File: chainer/function.py [hunk 10]
```diff
@@ -317,3 +339,179 @@ def unchain(self):
             if y_ref is not None:
                 y_ref.creator = None
         self.inputs = None
+
+    def add_hook(self, hook, name=None):
+        """Registers the function hook.
+
+        Args:
+            hook(~chainer.functions.FunctionHook):
```
Comment (unnonouno) [inline chainer/function.py]: `chaier.function.FunctionHook` ?

File: chainer/function.py [hunk 11]
```diff
@@ -317,3 +339,179 @@ def unchain(self):
...
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        del chainer.get_function_hooks()[self.name]
+
+    # forward
+    def forward_preprocess(self, function, in_data):
+        """Callback function invoked before forward propagation.
+
+        Args:
+            function(~chainer.function.Function): Function object to which
```
Comment (unnonouno) [inline chainer/function.py]: `chainer.Function` ?

File: chainer/function.py [hunk 12]
```diff
@@ -317,3 +339,179 @@ def unchain(self):
...
+            in_data(tuple of numpy.ndarray or tuple of cupy.ndarray):
+                Input data of forward propagation.
+        """
+        pass
+
+    def forward_postprocess(self, function, in_data):
+        """Callback function invoked after forward propagation.
+
+        Args:
+            function(~chainer.function.Function): Function object to which
```
Comment (unnonouno) [inline chainer/function.py]: `chainer.Function` ?

File: chainer/function.py [hunk 13]
```diff
@@ -317,3 +339,179 @@ def unchain(self):
...
+                Input data of forward propagation.
+        """
+        pass
+
+    # backward
+    def backward_preprocess(self, function, in_data, out_grad):
+        """Callback function invoked before backward propagation.
+
+        Args:
+            function(~chainer.function.Function): Function object to which
```
Comment (unnonouno) [inline chainer/function.py]: `chainer.Function`?

File: chainer/function.py [hunk 14]
```diff
@@ -317,3 +339,179 @@ def unchain(self):
...
+            out_grad(tuple of numpy.ndarray or tuple of cupy.ndarray):
+                Gradient data of backward propagation.
+        """
+        pass
+
+    def backward_postprocess(self, function, in_data, out_grad):
+        """Callback function invoked after backward propagation.
+
+        Args:
+            function(~chainer.function.Function): Function object to which
```
Comment (unnonouno) [inline chainer/function.py]: `chainer.Function` ?

File: chainer/function_hooks/timer.py [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+        assert xp == self.xp
+        self._postprocess(function)
+
+    def backward_postprocess(self, function, in_data, out_grad):
+        xp = cuda.get_array_module(*(in_data + out_grad))
+        assert xp == self.xp
+        self._postprocess(function)
+
+    def total_time(self):
+        """Returns total elapsed time."""
```
Comment (unnonouno) [inline chainer/function_hooks/timer.py]: Write return type of this method. `float` ?

**Comment by unnonouno** [comment]
OK. LGTM!

**Pull request merged**